src/Robot/MathematicianMonad.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE DeriveGeneric #-}
+
+module Robot.MathematicianMonad where
+
+import Robot.AutomationData
+import Robot.History
+import Robot.TableauFoundation
+
+import Control.Monad.State
+import GHC.Generics
+import Data.Aeson (FromJSON, ToJSON)
+
+-- | The state we have in the monad is called MathematicianState
+data MathematicianState = MathematicianState
+    { returnFreshName :: Int,
+      -- ^ An integer larger than all the names of free variables.
+      -- ready to be used as a new name
+      returnAutData :: AutData,
+      returnHistory :: History }
+      deriving (Show, Read, Generic)
+
+instance ToJSON MathematicianState
+instance FromJSON MathematicianState
+
+-- | The Mathematician monad combines State and Maybe capabilities
+type Mathematician = StateT MathematicianState Maybe
+
+-- | Lifting function mainly used to turn Maybes into Mathematicians
+liftMaybe :: (MonadPlus m) => Maybe a -> m a
+liftMaybe (Just x) = return x
+liftMaybe Nothing = mzero
+
+-- | Get a fresh name and increment the counter
+getFreshName :: Mathematician Int
```

Comment:
I think this function (and almost all the ones from here down) can be generalised to
```
getFreshName :: MonadState MathematicianState m => m Int
```
etc, meaning they'll work for the Mathematician monad but also for any other monad which has StateT 
MathematicianState in there somewhere